lib/ecto/adapters/postgres/connection.ex

```diff
@@ -925,18 +940,21 @@ if Code.ensure_loaded?(Postgrex) do
       :binary.replace(value, "'", "''", [:global])
     end
 
-    defp ecto_to_db({:array, t}),     do: [ecto_to_db(t), ?[, ?]]
-    defp ecto_to_db(:id),             do: "integer"
-    defp ecto_to_db(:serial),         do: "serial"
-    defp ecto_to_db(:bigserial),      do: "bigserial"
-    defp ecto_to_db(:binary_id),      do: "uuid"
-    defp ecto_to_db(:string),         do: "varchar"
-    defp ecto_to_db(:binary),         do: "bytea"
-    defp ecto_to_db(:map),            do: Application.fetch_env!(:ecto, :postgres_map_type)
-    defp ecto_to_db({:map, _}),       do: Application.fetch_env!(:ecto, :postgres_map_type)
-    defp ecto_to_db(:utc_datetime),   do: "timestamp"
-    defp ecto_to_db(:naive_datetime), do: "timestamp"
-    defp ecto_to_db(other),           do: Atom.to_string(other)
+    defp ecto_to_db({:array, t}),          do: [ecto_to_db(t), ?[, ?]]
+    defp ecto_to_db(:id),                  do: "integer"
+    defp ecto_to_db(:serial),              do: "serial"
+    defp ecto_to_db(:bigserial),           do: "bigserial"
+    defp ecto_to_db(:binary_id),           do: "uuid"
+    defp ecto_to_db(:string),              do: "varchar"
+    defp ecto_to_db(:binary),              do: "bytea"
+    defp ecto_to_db(:map),                 do: Application.fetch_env!(:ecto, :postgres_map_type)
+    defp ecto_to_db({:map, _}),            do: Application.fetch_env!(:ecto, :postgres_map_type)
+    defp ecto_to_db(:time_usec),           do: "time"
+    defp ecto_to_db(:utc_datetime),        do: "timestamp"
+    defp ecto_to_db(:utc_datetime_usec),   do: "timestamp"
+    defp ecto_to_db(:naive_datetime),      do: "timestamp"
+    defp ecto_to_db(:naive_datetime_usec), do: "timestamp"
```

Comment:
If we return `timestamp(6)` here, do we run into issues?

Reply:
Having this just be `"timestamp"` here allows the precision option to be set. By default, Postgres has no explicit bound on precision. Because of this it will work for a `_usec` type without having to set the precision to `6` like we do for mysql.

Reply:
The "issue" is that we can also do something like `from "posts", select: type(p.published_at, :naive_datetime_utc)` and we want that to have precision of 6 and we want the `:naive_datetime` to have precision of 0. So maybe ecto_to_db should always return the precision and column_type should handle when the precision is duplicated, something like:

```elixir
defp column_type(type, opts) when type in ~w(time time_usec)a do
  if precision = Keyword.get(opts, :precision) do
    ["time", ?(, to_string(precision), ?)]
  else
    ecto_to_db(type)
  end
end

defp column_type(type, opts) when type in ~w(naive_datetime utc_datetime utc_datetime_usec naive_datetime_usec)a do
  if precision = Keyword.get(opts, :precision) do
    ["timestamp", ?(, to_string(precision), ?)]
  else
    ecto_to_db(type)
  end
end

defp ecto_to_db(:time), do: "time(0)"
defp ecto_to_db(:time_usec), do: "time(6)"
# ...
```

Reply:
Actually, after further thoughts, I believe your code is right since we remove the precision on the Elixir side.

Reply:
Actually, your current code is correct, apologies. :)

Reply:
Hmm, I've been looking into your first response a bit deeper.

First of all, thanks for the explanation of the issue! That helped me look into something I missed — that `tagged_to_db/1` calls `ecto_to_db/1`. I didn't see that the first time through for some reason.

Wondering if you could help me understand the example you gave with this because I was actually persuaded by your first comment that this does need to be different. haha

I'm reading the following function in `Postgres.Connection` as building part of query that is actually sent to the DB. 
```
    defp expr(%Ecto.Query.Tagged{value: other, type: type}, sources, query) do
      [expr(other, sources, query), ?:, ?: | tagged_to_db(type)]
    end
```

So if we have a query  
`from "posts", select: type(p.published_at, :naive_datetime)`
with the changes in this PR it would build
`SELECT p0."published_at"::timestamp FROM "posts" AS p0`
but with a when using time, naive_datetime, or utc_datetime wouldn't the user be wanting the microseconds dropped if they exist?
`SELECT p0."published_at"::timestamp(0) FROM "posts" AS p0`

That's what I thought you were referring to anyway.   :)

Reply:
Currently I'm thinking if I change the column_type function for non-usec types to be

```
    defp column_type(type, _opts) when type in ~w(time utc_datetime naive_datetime)a,
      do: ecto_to_db(type)
```

and adjust the `ecto_to_db` functions for not-usec types to include `(0)` that it would cover all of the scenarios.

```
    defp ecto_to_db(:time),                do: "time(0)"
    defp ecto_to_db(:time_usec),           do: "time"
    defp ecto_to_db(:utc_datetime),        do: "timestamp(0)"
    defp ecto_to_db(:utc_datetime_usec),   do: "timestamp"
    defp ecto_to_db(:naive_datetime),      do: "timestamp(0)"
    defp ecto_to_db(:naive_datetime_usec), do: "timestamp"
```

Thoughts?

Whatever way we go I'll definitely also add some tests for these `Query.type/2` scenarios!

Reply:
> That's what I thought you were referring to anyway. :)

Yes, that's what I was referring to but then I realized it doesn't matter much because if we get more precision, we will just go ahead and discard it on the Ecto side. So let's definitely leave it as is. :)

Reply:
Ah, that's right! Gets pushed through the cast anyway. Thanks!